File: fsharp/accumulate/Accumulate.fs
module Accumulate

    let private cons (head:'a) (tail:'a list) = head::tail

    let rec private builder (differenceList:'b list -> 'b list) (func: 'a -> 'b) (list:'a list):'b list =
        match list with
        | [] -> differenceList []
        | (head: 'a)::(tail: 'a list) -> builder (differenceList << cons (func head)) func tail

    let accumulate<'a, 'b> (func: 'a -> 'b) (input: 'a list): 'b list =
        input
        |> builder id func

File: solutions/scala/accumulate/1/src/main/scala/Accumulate.scala
class Accumulate
{

  private def builder[A, B](stack:List[B], func: A => B, list:List[A]):List[B] =
    list match
    {
      case Nil => stack.reverse
      case (head: A)::(tail: List[A]) => builder(func(head)::stack, func,tail)
    }

  def accumulate[A, B](f: (A) => B, list : List[A]): List[B] = builder(Nil, f, list)
}
